[#4163] feat(trino-connector)  Support create user customized Trino connectors

diqiu50 opened the pull request:
### What changes were proposed in this pull request?

Make the CatalogConnectorFactory pluggable to support customized Trino connectors 

### Why are the changes needed?

Fix: #4163 

### Does this PR introduce _any_ user-facing change?

NO

### How was this patch tested?

Manually Test

yuqi1129 commented on trino-connector/src/main/java/org/apache/gravitino/trino/connector/GravitinoConfig.java:
```diff
@@ -69,6 +69,13 @@ public class GravitinoConfig {
+  private static final ConfigEntry GRAVITINO_CLUSTER =
+      new ConfigEntry(
+          "gravitino.cluster",
+          "The cluster name of the trino, if it's empty, all the catalogs handled by locally",
```
`all the catalogs handled by locally` ->  `all catalogs will be handled locally`

yuqi1129 commented on trino-connector/src/main/java/org/apache/gravitino/trino/connector/metadata/GravitinoCatalog.java:
```diff
@@ -116,4 +128,16 @@ public static String toJson(GravitinoCatalog catalog) throws JsonProcessingExcep
+    return properties.getOrDefault("cluster", "");
+  }
+
+  public boolean isLocally(String cluster) {
```
Locally is an adverb word, maybe you need to change it to `local`

yuqi1129 commented on trino-connector/src/main/java/org/apache/gravitino/trino/connector/catalog/jdbc/trino/TrinoClusterConnectorAdapter.java:
```diff
@@ -0,0 +1,90 @@
+  @Override
+  public Map<String, String> buildInternalConnectorConfig(GravitinoCatalog catalog) {
+    Map<String, String> config = new HashMap<>();
+    String jdbc_url = catalog.getProperty(TRINO_CLUSTER_URL_KEY, "");
```
`jdbc_url` is not a preferable Java naming style.

yuqi1129 commented on trino-connector/src/main/java/org/apache/gravitino/trino/connector/metadata/GravitinoCatalog.java:
```diff
@@ -116,4 +128,16 @@ public static String toJson(GravitinoCatalog catalog) throws JsonProcessingExcep
+  }
+
+  public boolean isLocally(String cluster) {
+    // Locally catalog means the gravitino connector has not config the cluster or
```
config -> configured

mchades commented on integration-test/src/test/java/org/apache/gravitino/integration/test/trino/TrinoQueryTestTool.java:
```diff
@@ -47,6 +47,10 @@ public static void main(String[] args) throws Exception {
+      options.addOption(
+          "test_host",
+          true,
+          "host for test server, if --auto is set to 'all', this option is ignored");
```
what's the meaning of `test server`?

diqiu50 replied:
The services server for run the tester. eg: hive, mysql, trino, gravitino server ...

yuqi1129 replied:
`host for test server` is very vague, can you give some examples about the possible value? besides `host for test server` means the the `type` of server or `address` of the server?

mchades commented on trino-connector/src/main/java/org/apache/gravitino/trino/connector/GravitinoConfig.java:
```diff
@@ -69,6 +69,13 @@ public class GravitinoConfig {
+  private static final ConfigEntry GRAVITINO_CLUSTER =
+      new ConfigEntry(
+          "gravitino.cluster",
+          "The cluster name of the trino, if it's empty, all catalogs will be handled locally",
```
It seems strange that `gravitino.cluster` is referred to as `The cluster name of the trino`.
BTW, Where should the user obtain the cluster name?

diqiu50 replied:
This will need to be considered by our Gravitino enterprise product in the future. For now, let's use the cluster name with the catalog's properties

yuqi1129 replied:
Can you provide an example here?

diqiu50 replied:
We can create a catalog with a cluster information:
```
call gravitino.system.create_catalog(
    'gt_mysql',
    'jdbc-mysql',
    map(
        array['jdbc-url', 'jdbc-user', 'jdbc-password', 'jdbc-driver', 'cluster', 'cluster.connection-url', 'cluster.connection-user'],
        array['jdbc:mysql://mysql/?useSSL=false', 'trino', 'ds123', 'com.mysql.cj.jdbc.Driver', 'c2', 'jdbc:trino://trino-remote:8080', 'admin']
    )
);
```

When the gravitino-trino-connector config like:
```
#Thu Jun 13 17:50:55 CST 2024
connector.name=gravitino
gravitino.metalake=test
gravitino.uri=http\://gt01.orb.local\:8090
trino.catalog.store=etc/catalog
trino.jdbc.uri=jdbc\:trino\://localhost\:8080
gravitino.cluster=c2
```

If the catalog's cluster name equals then the gravition-trino-connector 's  cluster configuration. The catalog is handle by local

yuqi1129 commented on trino-connector/src/main/java/org/apache/gravitino/trino/connector/metadata/GravitinoCatalog.java:
```diff
@@ -116,4 +128,16 @@ public static String toJson(GravitinoCatalog catalog) throws JsonProcessingExcep
+    return properties.getOrDefault("cluster", "");
+  }
+
+  public boolean isLocal(String cluster) {
```
isLocalCluster

diqiu50 replied:
We can create a catalog with a cluster information:
```
call gravitino.system.create_catalog(
    'gt_mysql',
    'jdbc-mysql',
    map(
        array['jdbc-url', 'jdbc-user', 'jdbc-password', 'jdbc-driver', 'cluster', 'cluster.connection-url', 'cluster.connection-user'],
        array['jdbc:mysql://mysql/?useSSL=false', 'trino', 'ds123', 'com.mysql.cj.jdbc.Driver', 'c2', 'jdbc:trino://trino-remote:8080', 'admin']
    )
);
```

When the gravitino-trino-connector config like:
```
#Thu Jun 13 17:50:55 CST 2024
connector.name=gravitino
gravitino.metalake=test
gravitino.uri=http\://gt01.orb.local\:8090
trino.catalog.store=etc/catalog
trino.jdbc.uri=jdbc\:trino\://localhost\:8080
gravitino.cluster=c2
```

If the catalog's cluster name equals then the gravition-trino-connector 's  cluster configuration. The catalog is handle by local

yuqi1129 commented on trino-connector/src/main/java/org/apache/gravitino/trino/connector/metadata/GravitinoCatalog.java:
```diff
@@ -116,4 +128,16 @@ public static String toJson(GravitinoCatalog catalog) throws JsonProcessingExcep
+  }
+
+  public boolean isLocal(String cluster) {
+    // Locally catalog means the gravitino connector has not configured the cluster or
```
Locally -> local, besides please move this comment to the method description.

mchades commented on integration-test/src/test/java/org/apache/gravitino/integration/test/trino/TrinoQueryTestTool.java:
```diff
@@ -47,6 +47,12 @@ public static void main(String[] args) throws Exception {
+      options.addOption(
+          "test_host",
+          true,
+          "host address for all test services (like gravitino server, trino, hive, postgresql, mysql..., "
+              + "if you services not running on one host. you need to set the arguments like --gravitino_uri=xxx and --trino_uri=xxx), "
+              + "default services address is 127.0.0.1, if --auto is set to 'all', this option is ignored");
```
suggestions: 
- Host address for all test services (including .........
- if the services are not running on the same host, set arguments like --gravitino_uri=xxx and --trino_uri=xxx),
- all test services will use 127.0.0.1 as a default host. If --auto is set to 'all', this option is ignored

mchades commented on trino-connector/src/main/java/org/apache/gravitino/trino/connector/GravitinoConfig.java:
```diff
@@ -69,6 +69,15 @@ public class GravitinoConfig {
 
+  private static final ConfigEntry GRAVITINO_CLUSTER =
+      new ConfigEntry(
+          "gravitino.cluster",
```
How about changing the name to `gravitino.trino_cluster`?

diqiu50 replied:
i think the cluster is not only for trino

mchades replied:
If so, how about `gravitino.query_cluster`?  Because the `gravitino.cluster` is easily mistaken for gravitino cluster, but currently gravitino does not have a clustering mode.

diqiu50 replied:
Replace to `cluster.region-code`

mchades commented on trino-connector/src/main/java/org/apache/gravitino/trino/connector/GravitinoConfig.java:
```diff
@@ -69,6 +69,15 @@ public class GravitinoConfig {
+  private static final ConfigEntry GRAVITINO_CLUSTER =
+      new ConfigEntry(
+          "gravitino.cluster",
+          "The cluster name of the Trino cluster, if it's empty, all the queries belongs the catalogs will be handled by the "
+              + "current Trino cluster, and if the cluster name is not equal to the catalog's cluster, the queries in the catalog "
+              + "will be handle by a remote Trino cluster",
```
If the user sets the cluster name as "my_test_trino," how does the connector determine the corresponding URI for "my_test_trino"?

diqiu50 replied:
The connector will check the url when the connector create. make sure the url is valid.

mchades replied:
You're missing the point, my question is, if a user configures `gravitino.cluster`, how should the URL corresponding to this cluster be configured?

diqiu50 replied:
Not sure. the configuration is just a cluster name.

yuqi1129 commented on trino-connector/src/main/java/org/apache/gravitino/trino/connector/catalog/CatalogConnectorFactory.java:
```diff
@@ -35,8 +36,11 @@ public class CatalogConnectorFactory {
   private static final Logger LOG = LoggerFactory.getLogger(CatalogConnectorFactory.class);
 
   private final HashMap<String, CatalogConnectorContext.Builder> catalogBuilders = new HashMap<>();
+  private final String cluster;
```
I would suggest you change it to `clusterName`?

mchades commented on integration-test/src/test/java/org/apache/gravitino/integration/test/trino/TrinoQueryTestTool.java:
```diff
@@ -47,6 +47,12 @@ public static void main(String[] args) throws Exception {
+      options.addOption(
+          "test_host",
+          true,
+          "host address for all test services (include gravitino server, trino, hive, postgresql, mysql..., "
```
Please capitalize the first letter.

diqiu50 commented:
This PR includes two changes:

1. Added a cloud-region configuration option to the Trino connector to identify which Gravitino cloud-region it belongs to. A Gravitino cloud-region refers to catalogs in the same data center and the corresponding execution clusters for Trino or Spark.
2. Implemented the logic for creating cascading query connectors in the Trino connector.

@jerryshao @mchades @yuqi1129 @FANNG1 please check if this part needs to be moved to a private repository.

yuqi1129 commented on trino-connector/src/main/java/org/apache/gravitino/trino/connector/GravitinoConfig.java:
```diff
@@ -69,6 +69,20 @@ public class GravitinoConfig {
+  private static final ConfigEntry GRAVITINO_CATALOG_CONNECTOR_FACTORY_CLASS_NAME =
+      new ConfigEntry(
+          "gravitino.catalog.connector.factory.class.name",
+          "The class name for the special CatalogConnectorFactory.",
```
for the special ->  custom

yuqi1129 replied:
Please add the following contents.

The specific class should implement interface `CatalogConnectorFactory`

yuqi1129 commented on trino-connector/src/main/java/org/apache/gravitino/trino/connector/GravitinoConnectorFactory.java:
```diff
@@ -120,4 +124,20 @@ public Connector create(
     return () -> null;
   }
+
+  CatalogConnectorFactory createCatalogConnectorFactory(GravitinoConfig config) {
```
Please add some comments on this method and highlight the point that we can support custom catalog connector factory.

yuqi1129 reviewed (approved):
LGTM

jerqi commented:
Maybe a typo `costomlize` -> `customize`
